README.md
<div>
<h2><a id="readme-general">OASIS Open Repository: cti-taxii-server</a></h2>

<p>This GitHub public repository ( <b><a href="https://github.com/oasis-open/cti-taxii-server">https://github.com/oasis-open/cti-taxii-server</a></b> ) was created at the request of the <a href="https://www.oasis-open.org/committees/cti/">OASIS Cyber Threat Intelligence (CTI) TC</a> as an <a href="https://www.oasis-open.org/resources/open-repositories/">OASIS Open Repository</a> to support development of open source resources related to Technical Committee work.</p>

<p>While this Open Repository remains associated with the sponsor TC, its development priorities, leadership, intellectual property terms, participation rules, and other matters of governance are <a href="https://github.com/oasis-open/cti-taxii-server/blob/master/CONTRIBUTING.md#governance-distinct-from-oasis-tc-process">separate and distinct</a> from the OASIS TC Process and related policies.</p>

<p>All contributions made to this Open Repository are subject to open source license terms expressed in the <a href="https://www.oasis-open.org/sites/www.oasis-open.org/files/BSD-3-Clause.txt">BSD-3-Clause License</a>.  That license was selected as the declared <a href="https://www.oasis-open.org/resources/open-repositories/licenses">"Applicable License"</a> when the Open Repository was created.</p>

<p>As documented in <a href="https://github.com/oasis-open/cti-taxii-server/blob/master/CONTRIBUTING.md#public-participation-invited">"Public Participation Invited</a>", contributions to this OASIS Open Repository are invited from all parties, whether affiliated with OASIS or not.  Participants must have a GitHub account, but no fees or OASIS membership obligations are required.  Participation is expected to be consistent with the <a href="https://www.oasis-open.org/policies-guidelines/open-repositories">OASIS Open Repository Guidelines and Procedures</a>, the open source <a href="https://github.com/oasis-open/cti-taxii-server/blob/master/LICENSE">LICENSE</a> designated for this particular repository, and the requirement for an <a href="https://www.oasis-open.org/resources/open-repositories/cla/individual-cla">Individual Contributor License Agreement</a> that governs intellectual property.</p>

</div>

<div>
<h2><a id="purposeStatement">Statement of Purpose</a></h2>


<p>Statement of Purpose for this OASIS Open Repository (cti-taxii-server) as <a href="https://lists.oasis-open.org/archives/cti/201707/msg00039.html">proposed</a> and <a href="https://lists.oasis-open.org/archives/cti/201708/msg00000.html">approved</a> [<a href="https://issues.oasis-open.org/browse/TCADMIN-2640">bis</a>] by the TC:</p>

<p>The taxii-server code under development in this GitHub repository is a Python library which will process HTTPS requests from TAXII clients in conformance with the <a href="http://docs.oasis-open.org/cti/taxii/">TAXII specification</a>.</p>

</div>

<div><h2><a id="purposeClarifications">Additions to Statement of Purpose</a></h2>

<p>Repository Maintainers may include here any clarifications &mdash; any additional sections, subsections, and paragraphs that the Maintainer(s) wish to add as descriptive text, reflecting (sub-) project status, milestones, releases, modifications to statement of purpose, etc.  The project Maintainers will create and maintain this content on behalf of the participants.</p>
</div>

<div>
<h2><a id="maintainers">Maintainers</a></h2>

<p>Open Repository <a href="https://www.oasis-open.org/resources/open-repositories/maintainers-guide">Maintainers</a> are responsible for oversight of this project's community development activities, including evaluation of GitHub <a href="https://github.com/oasis-open/cti-taxii-server/blob/master/CONTRIBUTING.md#fork-and-pull-collaboration-model">pull requests</a> and <a href="https://www.oasis-open.org/policies-guidelines/open-repositories#repositoryManagement">preserving</a> open source principles of openness and fairness. Maintainers are recognized and trusted experts who serve to implement community goals and consensus design preferences.</p>

<p>Initially, the associated TC members have designated one or more persons to serve as Maintainer(s); subsequently, participating community members may select additional or substitute Maintainers, per <a href="https://www.oasis-open.org/resources/open-repositories/maintainers-guide#additionalMaintainers">consensus agreements</a>.</p>

<p><b><a id="currentMaintainers">Current Maintainers of this Open Repository</a></b></p>

<ul>
<li><a href="mailto:gback@mitre.org">Greg Back</a>; GitHub ID: <a href="https://github.com/gtback/">https://github.com/gtback/</a>; WWW: <a href="https://www.mitre.org/">MITRE Corporation</a></li>
<li><a href="mailto:rpiazza@mitre.org">Rich Piazza</a>; GitHub ID: <a href="https://github.com/rpiazza/">https://github.com/rpiazza/</a>; WWW: <a href="https://www.mitre.org/">MITRE Corporation</a></li>
</ul>


</div>

<div><h2><a id="aboutOpenRepos">About OASIS Open Repositories</a></h2>

<p><ul>
<li><a href="https://www.oasis-open.org/resources/open-repositories/">Open Repositories: Overview and Resources</a></li>
<li><a href="https://www.oasis-open.org/resources/open-repositories/faq">Frequently Asked Questions</a></li>
<li><a href="https://www.oasis-open.org/resources/open-repositories/licenses">Open Source Licenses</a></li>
<li><a href="https://www.oasis-open.org/resources/open-repositories/cla">Contributor License Agreements (CLAs)</a></li>
<li><a href="https://www.oasis-open.org/resources/open-repositories/maintainers-guide">Maintainers' Guidelines and Agreement</a></li>
</ul></p>

</div>

<div><h2><a id="feedback">Feedback</a></h2>

<p>Questions or comments about this Open Repository's activities should be composed as GitHub issues or comments. If use of an issue/comment is not possible or appropriate, questions may be directed by email to the Maintainer(s) <a href="#currentMaintainers">listed above</a>.  Please send general questions about Open Repository participation to OASIS Staff at <a href="mailto:repository-admin@oasis-open.org">repository-admin@oasis-open.org</a> and any specific CLA-related questions to <a href="mailto:repository-cla@oasis-open.org">repository-cla@oasis-open.org</a>.</p>

</div>


CHANGES.txt
Version 3.0.0
2020-07-07
Major release to support TAXII 2.1
- #116 MemoryBackend enhancement to allow spec_version auto filtering
- #117 MongoDB document structure and pipeline change (breaking change)
- #119 Update test suite to prevent errors in MongoDB causing inconsistent behavior in requests
- #120 MemoryBackend update to default filter behavior when not provided in request
- #126 Fix stability issues with inconsistent behavior in requests (mainly affecting old Python versions)
- #127 Dropped support for Python 2.7, and 3.4 and Travis CI testing

Version 2.0.1
2020-04-03
- #112 Test revamp for consistency across backends
- Resolves an issue with MongoBackend delete object where wrong version would be removed
- Resolved an issue with Content-Type header validation causing 415 Response for valid content

Version 2.0.0
2020-03-27
First major and stable release to support TAXII 2.1
- Supports pagination
- Supports all defined match filters (id, spec_version, version, type)
- Header Validation for GET, POST requests
- Updated tests (except view-level tests)

Version 1.1.0
2020-07-02
Patch release to fix issues on TAXII 2.0
- #129 Dropped Python versions older than 3.5
- #128 The Range header was incorrectly checked/enforced according to RFC 7233 (@dougle)

Version 1.0.0
2020-03-26
First major and stable release to support TAXII 2.0
- #108, #109 Additional checks to Accept, Content-Type and Content-Length headers and API root existence
- #108 Add or clean documentation for Endpoints defined in the spec
- #108 Design change in validation procedures. These no longer used to evaluate True or raise Exception
- #108 Update packaging, tests per new changes, code quality
- #108 Bug under MemoryBackend that would cause a 400 Error when filtering under manifests

Version 0.5.0
2020-03-20
- #60 Fixes to tests and Default Values
- #58, #91,  Updating documentation, or configuration
- #59 Directory Backend Plus Auth Improvements
- #74, #104, #105 Update docker configuration
- #77 mongo filters, example data, port methods that overlap with 2.1
- #76, #78 Directory Backend changes and general improvements, checks
- The MongoBackend has received changes in data representation, bug fixes, and filters.
- The MemoryBacked has received changes in data representation, bug fixes, and filters.
- JWT login support, API Key, heartbeat, and routes functionality via Endpoints.
- Headers are being checked.

Version 0.4.0
2019-07-16
- @jamesfitzsimons implemented Pagination support and fixed a couple of MongoDB errors
- #47, #51 Fix issue when objects returned from MongoBackend would strip some properties out
- #28 Added support for X-TAXII Custom Headers
- Added new Docker & Docker Compose configuration:
- Uses mongo image, mongo-express (lightweight web administrative interface for mongoDB) and the medallion build which is slightly tweaked
- The following bugs were fixed:
- The MemoryBackend was manipulating its manifest data. When using match[version] against a manifest resource with multiple versions available. It would get modified permanently
- Removed MongoBackend $project in pipelines that would cause loss of properties or custom properties
- Errors produced by the TAXII Server now return JSON that align with Section 3.6 of the spec
- Simplified some of the code for checking permission or determining if a collection exists
- The rest are code quality changes and consistency

Version 0.3.1
2018-10-05
- Fix packaging problem

Version 0.3.0
2018-10-05
- #41 Fixed MemoryBackend bug where checking for an object was incorrectly being performed.
- #34 Performance enhancements for the MongoDBBackend.
- #29 #30 #31 #30 Improvements to the server error handling.

Version 0.2.1
2018-03-26
- Fix packaging problem

Version 0.2.0
2018-03-26
- Fix error in MongoBackend that resulted in manifests not being updated
- Fix error in MongoBackend that resulted in manifests versions not being sorted properly
- Use argparse for medallion CLI
- Return correct status responses from MemoryBackend and MongoBackend
- Decouple some medallion initialization to allow usage as library or further integration
- Use application proxy for endpoints
- Allow for dynamic backend via JSON config file or set methods
- Add logging messages
- Improve code coverage/testing

Version 0.1.0
2017-07-14
- Initial release


medallion/views/objects.py
import re

from flask import Blueprint, Response, current_app, json, request

from . import (
    MEDIA_TYPE_STIX_V20, MEDIA_TYPE_TAXII_V20,
    validate_stix_version_parameter_in_accept_header,
    validate_taxii_version_parameter_in_accept_header
)
from .. import auth
from ..common import get_timestamp
from ..exceptions import ProcessingError
from .discovery import api_root_exists

objects_bp = Blueprint("objects", __name__)


def permission_to_read(api_root, collection_id):
    collection_info = current_app.medallion_backend.get_collection(api_root, collection_id)
    if collection_info["can_read"] is False:
        raise ProcessingError("Forbidden to read collection '{}'".format(collection_id), 403)


def permission_to_write(api_root, collection_id):
    collection_info = current_app.medallion_backend.get_collection(api_root, collection_id)
    if collection_info["can_write"] is False:
        raise ProcessingError("Forbidden to write collection '{}'".format(collection_id), 403)


def collection_exists(api_root, collection_id):
    if not current_app.medallion_backend.get_collection(api_root, collection_id):
        raise ProcessingError("Collection '{}' not found".format(collection_id), 404)


def validate_version_parameter_in_content_type_header():
    content_type = request.headers.get("content_type", "").replace(" ", "").split(",")
    found = False

    for item in content_type:
        result = re.match(r"^application/vnd\.oasis\.stix\+json(;version=(\d\.\d))?$", item)
        if result:
            if len(result.groups()) >= 2:
                version_str = result.group(2)
                if version_str != "2.0":  # The server only supports 2.0 at the moment
                    raise ProcessingError("The server does not support version {}".format(version_str), 415)
            found = True
            break

    if found is False:
        raise ProcessingError("Media type in the Content-Type header is invalid or not found", 415)


def validate_size_in_request_body(api_root):
    api_root = current_app.medallion_backend.get_api_root_information(api_root)
    max_length = api_root["max_content_length"]
    try:
        content_length = int(request.headers.get("content_length", ""))
    except ValueError:
        raise ProcessingError("The server did not understand the request or headers", 400)
    if content_length > max_length or content_length <= 0:
        raise ProcessingError("Content-Length header not valid or exceeds maximum!", 413)


def get_range_request_from_headers():
    if request.headers.get("Range") is not None:
        try:
            matches = re.match(r"^items=(\d+)-(\d+)$", request.headers.get("Range"))
            start_index = int(matches.group(1))
            end_index = int(matches.group(2))
            # check that the requested number of items isn't larger than the maximum support server page size
            # the +1 and -1 below account for the fact that paging is zero index based.
            if (end_index - start_index) + 1 > current_app.taxii_config["max_page_size"]:
                end_index = start_index + (current_app.taxii_config["max_page_size"] - 1)
            return start_index, end_index
        except (AttributeError, ValueError) as e:
            raise ProcessingError("Bad Range header supplied", 400, e)
    else:
        return 0, current_app.taxii_config["max_page_size"] - 1


def get_custom_headers(headers, api_root, collection_id, start, end):
    try:
        manifest = current_app.medallion_backend.get_object_manifest(
            api_root, collection_id, request.args, ("id",),  start, end)[1]
        if manifest:
            times = sorted(map(lambda x: x["date_added"], manifest))

            if len(times) > 0:
                headers['X-TAXII-Date-Added-First'] = times[0]
                headers['X-TAXII-Date-Added-Last'] = times[-1]
    except Exception as e:
        raise ProcessingError("Unable to build response headers", 400, e)
    return headers


def get_response_status_and_headers(start_index, total_count, objects):
    # If the requested range is outside the size of the result set, return a HTTP 416
    if start_index >= total_count > 0:
        error = ProcessingError("The requested range is outside the size of the result set", 416)
        error.headers = {
            "Accept-Ranges": "items",
            "Content-Range": "items */{}".format(total_count),
        }
        raise error

    # If no range request was supplied, and we can return the whole result set in one go, then do so.
    if request.headers.get("Range") is None and total_count < current_app.taxii_config["max_page_size"]:
        status = 200
        headers = {"Accept-Ranges": "items"}
    else:
        # The minus one below is due to the fact that the range is zero based
        status = 206
        # This check is required as if the number of objects returned is zero, subtracting one would give us
        # an end index of -1
        if len(objects) == 0:
            end_index = start_index
        else:
            end_index = start_index + (len(objects) - 1)
        headers = {
            "Accept-Ranges": "items",
            "Content-Range": "items {}-{}/{}".format(start_index, end_index, total_count),
        }
    return status, headers


@objects_bp.route("/<string:api_root>/collections/<string:collection_id>/objects/", methods=["GET", "POST"])
@auth.login_required
def get_or_add_objects(api_root, collection_id):
    """
    Defines TAXII API - Collections:
    `Get Objects Section (5.3) <http://docs.oasis-open.org/cti/taxii/v2.0/cs01/taxii-v2.0-cs01.html#_Toc496542738>`__
    and `Add Objects Section (5.4) <http://docs.oasis-open.org/cti/taxii/v2.0/cs01/taxii-v2.0-cs01.html#_Toc496542739>`__.

    Args:
        api_root (str): the base URL of the API Root
        collection_id (str): the `identifier` of the Collection being requested

    Returns:
        bundle: GET -> A STIX 2.0 Bundle upon successful requests. Additional information
             `here <http://docs.oasis-open.org/cti/stix/v2.0/cs01/part1-stix-core/stix-v2.0-cs01-part1-stix-core.html#_Toc496709292>`__.
        status: POST -> An Status Resource upon successful requests. Additional information
             `here <http://docs.oasis-open.org/cti/taxii/v2.0/cs01/taxii-v2.0-cs01.html#_Toc496542732>`__.

    """
    # TODO: Check if user has access to read or write objects in collection - right now just check for permissions on the collection.
    request_time = get_timestamp()  # Can't I get this from the request itself?
    if request.method == "GET":
        validate_stix_version_parameter_in_accept_header()
        api_root_exists(api_root)
        collection_exists(api_root, collection_id)
        permission_to_read(api_root, collection_id)
        start_index, end_index = get_range_request_from_headers()
        total_count, objects = current_app.medallion_backend.get_objects(
            api_root, collection_id, request.args, ("id", "type", "version"), start_index, end_index,
        )
        if objects:
            status, headers = get_response_status_and_headers(start_index, total_count, objects["objects"])
            headers = get_custom_headers(headers, api_root, collection_id, start_index, end_index)
            return Response(
                response=json.dumps(objects),
                status=status,
                headers=headers,
                mimetype=MEDIA_TYPE_STIX_V20,
            )
        raise ProcessingError("Collection '{}' has no objects available".format(collection_id), 404)
    elif request.method == "POST":
        validate_taxii_version_parameter_in_accept_header()
        validate_version_parameter_in_content_type_header()
        api_root_exists(api_root)
        collection_exists(api_root, collection_id)
        permission_to_write(api_root, collection_id)
        validate_size_in_request_body(api_root)

        status = current_app.medallion_backend.add_objects(api_root, collection_id, request.get_json(force=True), request_time)
        return Response(
            response=json.dumps(status),
            status=202,
            mimetype=MEDIA_TYPE_TAXII_V20,
        )


@objects_bp.route("/<string:api_root>/collections/<string:collection_id>/objects/<string:object_id>/", methods=["GET"])
@auth.login_required
def get_object(api_root, collection_id, object_id):
    """
    Defines TAXII API - Collections:
    `Get Object Section (5.5) <http://docs.oasis-open.org/cti/taxii/v2.0/cs01/taxii-v2.0-cs01.html#_Toc496542740>`__

    Args:
        api_root (str): the base URL of the API Root
        collection_id (str): the `identifier` of the Collection being requested
        object_id (str): the `identifier` of the object being requested

    Returns:
        bundle: GET -> A STIX 2.0 Bundle upon successful requests. Additional information
             `here <http://docs.oasis-open.org/cti/stix/v2.0/cs01/part1-stix-core/stix-v2.0-cs01-part1-stix-core.html#_Toc496709292>`__.

    """
    # TODO: Check if user has access to objects in collection - right now just check for permissions on the collection
    validate_stix_version_parameter_in_accept_header()
    api_root_exists(api_root)
    collection_exists(api_root, collection_id)
    permission_to_read(api_root, collection_id)

    objects = current_app.medallion_backend.get_object(api_root, collection_id, object_id, request.args, ("version",))
    if objects:
        return Response(
            response=json.dumps(objects),
            status=200,
            mimetype=MEDIA_TYPE_STIX_V20,
        )
    raise ProcessingError("Object '{}' not found".format(object_id), 404)
